gax/src/main/java/com/google/api/gax/batching/BatcherStats.java

```diff
@@ -0,0 +1,194 @@
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.batching;
+
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutureCallback;
+import com.google.api.core.ApiFutures;
+import com.google.api.core.SettableApiFuture;
+import com.google.api.gax.rpc.ApiException;
+import com.google.api.gax.rpc.StatusCode;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import javax.annotation.Nullable;
+
+/**
+ * This class keeps the statistics about failed operations(both at RPC and ElementT level) in {@link
+ * Batcher}. This provides the count of individual exception failure and count of each failed {@link
+ * StatusCode.Code} occurred in the batching process.
+ */
+class BatcherStats {
+
+  private final Map<Class, Integer> requestExceptionCounts = new ConcurrentHashMap<>();
+  private final Map<StatusCode.Code, Integer> requestStatusCounts = new ConcurrentHashMap<>();
+  private final AtomicInteger partialBatchFailures = new AtomicInteger(0);
+  private final Map<Class, Integer> entryExceptionCounts = new ConcurrentHashMap<>();
+  private final Map<StatusCode.Code, Integer> entryStatusCounts = new ConcurrentHashMap<>();
+
+  private final Object lock = new Object();
+
+  synchronized void recordBatchFailure(Throwable throwable) {
+    Class exceptionClass = throwable.getClass();
+
+    synchronized (lock) {
+      if (throwable instanceof ApiException) {
+        StatusCode.Code code = ((ApiException) throwable).getStatusCode().getCode();
+        exceptionClass = ApiException.class;
+
+        if (requestStatusCounts.containsKey(code)) {
+          Integer codeCount = requestStatusCounts.get(code);
+          requestStatusCounts.put(code, ++codeCount);
+
+        } else {
+          requestStatusCounts.put(code, 1);
+        }
```

Comment:
I think this would be nicer:
```java
int oldCount = MoreObjects.firstNonNull(requestStatusCounts.get(code), 0);
requestStatusCounts.put(code, oldCount + 1);
```

Reply:
Thanks again, have updated this and everywhere else with this approach.